### ISSUE TITLE
Form elements in Global Styles - iteration for WP 7.2

### ISSUE THREAD

**Issue by MaggieCabrera** [WordPress/gutenberg#34198]
## Overall goal

Let users style common form elements consistently from Global Styles, without writing custom CSS or editing `theme.json` by hand.

The `button`, `textInput`, and `select` elements are already supported in `theme.json`. This iteration exposes that work in the editor, adds the missing label element, and improves support for interactive input states.

Prior art: #34198.

## Scope for this release

- [x] #81647
- [x] #81648
- [ ] #71382
- [ ] #76534
- [ ] #81649

## Nice to have

- [ ] #81650
- [ ] #65343
- [ ] #52615
- [ ] #57314
- [x] #68195
- [x] #61353

**Comment by oandregal** [comment]
For context, a related conversation is https://github.com/WordPress/gutenberg/issues/34141#issuecomment-902178472

**Comment by oandregal** [comment]
These elements make sense to me as cross-block components.

The only bit I'm unsure about is the button: we already have a button block. Ideally, all complex blocks would use it as an inner block - that will solve the issue. I understand that's not always possible, though. If we create a button element, how would that interact with the styles in the button block? Say we have this:

```json
{
  "styles": {
    "elements": {
      "button": {
        "color": {
          "text": "antiquewhite",
          "background": "coral"
        }
      }
    },
    "blocks": {
      "core/button": {
        "color": {
          "text": "deeppink"
        }
      }
    }
  }
}
```

What would be the expected CSS?

**Comment by MaggieCabrera** [comment]
> The only bit I'm unsure about is the button: we already have a button block. Ideally, all complex blocks would use it as an inner block - that will solve the issue. I understand that's not always possible, though. If we create a button element, how would that interact with the styles in the button block? 

I wish the button block could be used by other blocks as well, that way when the button block supports hover states, so will the rest of the blocks using it, right? As to the CSS on your example, I would expect it to be the same as the headings, where the styles defined for the block override the ones for the element.

**Comment by oandregal** [comment]
I'll let other people share their thoughts. It seems this could be useful, in my view. I'll have to see specific examples for the button, but that's something that can be explored in a specific PR with a theme that uses the core button and addresses buttons as elements.

**Comment by overclokk** [comment]
Also a `link` element could be a button (or at least styled like a button), so why not add a controller for the block to indicate a `type` of the element, the controller will add a css class to the block we can refer in the theme.json to apply the style, so we could have:

```json
{
  "styles": {
    "types": {
      "button": {
        "color": {
          "text": "antiquewhite",
          "background": "coral"
        }
      }
    },
    "blocks": {
      "core/button": {
        "color": {
          "text": "deeppink"
        }
      }
    }
  }
}
```

A CSS class name could be something like `is-type-button` or `has-type-button` or wethever.

This way we only have **one** global CSS class name for all button types, button block, file block, comment form block and search block, and we can activate a type button also for link in the navbar, in the content  and why not also for radio button, checkbox and input fields.

For nested element we can apply a few line of CSS to create a different design, let say for example for the search bar we want a design with "icon, input and button" grouped together.

This way we can create something similar to what the library like Bootstrap does.

**Comment by carolinan** [comment]
I believe that this has been solved for buttons? 
How can we move this forward and discuss the other form elements?

**Comment by MaggieCabrera** [comment]
> I believe that this has been solved for buttons? How can we move this forward and discuss the other form elements?

I think no one has picked anything up, but I don't think there's any blockers besides wanting to work on it? Working on the input or labels would be great, it would give themers and users much better control over the search block as a whole. Also, when we upgrade the comments form to a proper block it would be great to have them all be made with elements from the start.

**Comment by mtias** [comment]
All form element need a lot more thinking before being considered for the elements API, including how they are going to be controlled, and how they are going to be leveraged by block authors.

**Comment by senadir** [comment]
> I believe that this has been solved for buttons?

I'd love to see any docs around this?

**Comment by MaggieCabrera** [comment]
> > I believe that this has been solved for buttons?
> 
> I'd love to see any docs around this?

Elements are documented [here](https://developer.wordpress.org/block-editor/how-to-guides/themes/theme-json/#element-styles)

**Comment by dkjensen** [comment]
+1

**Comment by mikeritter** [comment]
This is holding me back.

I don't see a roadmap to keep this on track.

## [HTML Form elements](https://www.w3schools.com/html/html_form_elements.asp)

```html
<input>
<label>
<select>
<textarea>
<button>
<fieldset>
<legend>
<datalist>
<output>
<option>
<optgroup>
```

## [Inputs](https://www.w3schools.com/html/html_form_input_types.asp)?

```html
<input type="button">
<input type="checkbox">
<input type="color">
<input type="date">
<input type="datetime-local">
<input type="email">
<input type="file">
<input type="hidden">
<input type="image">
<input type="month">
<input type="number">
<input type="password">
<input type="radio">
<input type="range">
<input type="reset">
<input type="search">
<input type="submit">
<input type="tel">
<input type="text">
<input type="time">
<input type="url">
<input type="week">
```

## The select element

```html
<label for="cars">Choose a car:</label>
<select id="cars" name="cars">
  <option value="volvo">Volvo</option>
  <option value="saab">Saab</option>
  <option value="fiat">Fiat</option>
  <option value="audi">Audi</option>
</select>
```

## Fieldset

```html
  <fieldset>
    <legend>Personalia:</legend>
    <label for="fname">First name:</label><br>
    <input type="text" id="fname" name="fname" value="John"><br>
    <label for="lname">Last name:</label><br>
    <input type="text" id="lname" name="lname" value="Doe"><br><br>
    <input type="submit" value="Submit">
  </fieldset>
```

## [Optgroup](https://www.w3schools.com/tags/tag_optgroup.asp)

```html
<label for="cars">Choose a car:</label>
<select  name="cars" id="cars">
  <optgroup label="Swedish Cars">
    <option value="volvo">Volvo</option>
    <option value="saab">Saab</option>
  </optgroup>
  <optgroup label="German Cars">
    <option value="mercedes">Mercedes</option>
    <option value="audi">Audi</option>
  </optgroup>
</select>
```

**Comment by mikeritter** [comment]
> > I believe that this has been solved for buttons? How can we move this forward and discuss the other form elements?
> 
> I think no one has picked anything up, but I don't think there's any blockers besides wanting to work on it? Working on the **input** or **labels** would be great, it would give themers and users much better control over the **search block** as a whole. Also, when we upgrade the **comments form to a proper block** it would be great to have them all be made with elements from the start.

## Suggested Pathway

1. Identify sensible default properties to override for general `input` element (i.e., user agent defaults, common properties, states)
2. Identify layout patterns for grouping `label` and `input` elements (ul, div, label)
3. Repeat 1 for `label` then apply wrapper style to label
4. Apply to Search block
5. Apply to Comments block

**Comment by mikeritter** [comment]
I just found the [Jetpack Forms Block](https://jetpack.com/support/jetpack-blocks/contact-form/) with much of this heavy lifting built in. Here is a link to [the GitHub repo for Form block](https://github.com/Automattic/jetpack/tree/d4068d52c35a30edc01b9356a4764132aeb532fd/projects/plugins/jetpack/modules/contact-form).

**Comment by scruffian** [comment]
I have an idea for this here: https://github.com/WordPress/gutenberg/pull/51337

**Comment by carolinan** [comment]
Can this be closed as completed? Or can you add an update about what remains?

**Comment by MaggieCabrera** [comment]
> Can this be closed as completed? Or can you add an update about what remains?

I'm happy to close but I don't think we are covering all form elements. I am also not sure if we should! We certainly have covered the most used ones, we can leave the discussion open or close it and reopen if someone comes up with the need to revisit?

**Comment by scruffian** [comment]
Label seems like the obvious one that's missing, if someone wants to spin up a quick PR :)

**Comment by carolinan** [comment]
@t-hamano  Hi, wasn't the password protected form fixed in core in 7.1? Then we should close 61353.

**Comment by t-hamano** [comment]
> Hi, wasn't the password protected form fixed in core in 7.1? Then we should close 61353.

Yes, that should have been fixed by [Changeset 62754](https://core.trac.wordpress.org/changeset/62754).
